### PR TITLE
Algolia integration and Indexing the entities

### LINKED ISSUE #243: Install Algolia and index the entity Project and ProjectClassification
As described in #242 this issue aims to install Algolia in [LibreCores](https://www.librecores.org) and index the entity to make a better search experience.

### PR DESCRIPTION
**Pull request by sandipbhuyan**
Resolves #243. This PR is pointing to install Agolia and indexing the entities Project and ProjectClassification

### CODE REVIEW

File: ansible/roles/web/templates/parameters.yml.j2 [hunk 1]
```diff
@@ -45,3 +45,6 @@ parameters:
     # Register at https://sentry.io
     sentry_dsn:        '{{ site_sentry_dsn }}'
     sentry_env:        '{{ site_sentry_env }}'
+
+    env(ALGOLIA_APP_ID):  '{{ site_algolia_app_id }}'
+    env(ALGOLIA_API_KEY): '{{ site_algolia_api_key }}'
```
Comment (imphil) [inline ansible/roles/web/templates/parameters.yml.j2]: Please add a comment what these parameters are for and where to get them from, and end the file with the newline. (like all files, i.e. fix this across this PR).

File: ansible/secrets.dist/dev-vagrant.secrets.yml [hunk 2]
```diff
@@ -14,3 +14,7 @@ site_smtp_port: 'INSERT_SMTP_PORT_HERE'
 site_smtp_encryption: 'ssl' # options: tls, ssl, login
 site_smtp_user: 'INSERT_SMTP_USER_HERE'
 site_smtp_password: 'INSERT_SMTP_PASSWORD_HERE'
+
+# Algolia
+site_algolia_app_id: "INSERT_ALGOLIA_APP_ID_HERE"
+site_algolia_api_key: "INSERT_ALGOLIA_API_KEY_HERE"
```
Comment (imphil) [inline ansible/secrets.dist/dev-vagrant.secrets.yml]: Even though it does not change functionality, stick with single quotes (') for consistency. (Multiple times in this PR.)

File: ansible/vars/dev-vagrant.user.yml [hunk 3]
```diff
@@ -6,6 +6,7 @@
 
 
 # Require confirmation of the email address of newly registered users.
+# Require site_algolia_app_id and site_algolia_api_key of algolia application.
```
Comment (imphil) [inline ansible/vars/dev-vagrant.user.yml]: This looks strange within this mailer comment?

File: ansible/vars/dev-vagrant.user.yml [hunk 4]
```diff
@@ -6,6 +6,7 @@
...
 # Requires a working mailer setup. Add SMTP credentials in
 # secrets/dev-vagrant.secrets.yml before enabling this setting.
-#site_user_email_confirmation_enabled: true
+# site_user_email_confirmation_enabled: true
```
Comment (imphil) [inline ansible/vars/dev-vagrant.user.yml]: This comment is using intentionally no space after the "#" sign to show that this is not a comment, but a commented-out config setting. This is commonly done in example or preconfigured config files.

File: doc/src/devenv/tipsandtricks.rst [hunk 5]
```diff
@@ -110,3 +110,22 @@ Check the coding style of PHP code
...
+
+Algolia indices configuration
+---------------------------
+For configuring algolia in development environment first specify the Apllication ID and Admin API Key in ``parameter.yml``
+Clear the indices and import the settings from algolia
+
+Clear indices
+-------------
+.. code-block:: bash
+qqq
```
Comment (imphil) [inline doc/src/devenv/tipsandtricks.rst]: remove.

File: doc/src/devenv/tipsandtricks.rst [hunk 6]
```diff
@@ -110,3 +110,22 @@ Check the coding style of PHP code
   vm$> cd /var/www/lc/site
   vm$> ./vendor/bin/phpcs --runtime-set ignore_warnings_on_exit true -s \
     && echo You can commit: No errors found!
+
+Algolia indices configuration
+---------------------------
```
Comment (imphil) [inline doc/src/devenv/tipsandtricks.rst]: Please use "-" until the end of the headline.

File: doc/src/devenv/tipsandtricks.rst [hunk 7]
```diff
@@ -110,3 +110,22 @@ Check the coding style of PHP code
...
+For configuring algolia in development environment first specify the Apllication ID and Admin API Key in ``parameter.yml``
```
Comment (imphil) [inline doc/src/devenv/tipsandtricks.rst]: "To configure Algolia in the development environment you need to specify the Application ID (site_algolia_app_id) and the Admin API Key (site_algolia_api_key) in the ansible/secrets/dev-vagrant.secrets.yml file."

Is it really called "Admin API Key"?

File: doc/src/devenv/tipsandtricks.rst [hunk 8]
```diff
@@ -110,3 +110,22 @@ Check the coding style of PHP code
...
+Clear the indices and import the settings from algolia
```
Comment (imphil) [inline doc/src/devenv/tipsandtricks.rst]: You say here it imports the settings, below you say it imports the indices. What is correct?

File: site/app/config/config.yml [hunk 9]
```diff
@@ -199,3 +199,19 @@ sentry:
...
+
+# algolia search indices configuration
+algolia_search:
+  prefix: ''
+  indices:
+    - name: projects
+      class: Librecores\ProjectRepoBundle\Entity\Project
+
+    - name: classifications
+      class: Librecores\ProjectRepoBundle\Entity\ProjectClassification
```
Comment (imphil) [inline site/app/config/config.yml]: Is it really necessary to specify the classifications as separate index, doesn't this go into the projects index?

File: ansible/secrets.dist/dev-vagrant.secrets.yml (hunk 2 above)
Comment (sandipbhuyan) [reply ansible/secrets.dist/dev-vagrant.secrets.yml]: There were some (") quotes inside the same file should I change those too ??

File: doc/src/devenv/tipsandtricks.rst (hunk 7 above)
Comment (sandipbhuyan) [reply doc/src/devenv/tipsandtricks.rst]: Yes algolia use two keys one for inserting data into algolia(Admin API  Key)  another for searching functionality (Search-Only API Key)

**Comment by imphil** [comment]
I was going ahead and merge that but now that @agathver has brought up some issues in https://github.com/librecores/librecores-web/pull/246 which we might want to get resolved first in this PR (since the other once just folds in the changes of this one, and needs to be rebased on top of this one after it goes in).

- I agree with Amitosh about the missing newline at the end of the file. Look through the GitHub diff and the markers at the end of the files when a newline is missing, and add those.
- I also agree about the `@inheritdoc` usage.
- I'm not sure about what is meant by the comment that `env(ALGOLIA_APP_ID):` and friends are not required. Isn't that the (recommended/documented) way to configure Algolia with this version of Symfony? Somewhere this information is needed, isn't it?

File: ansible/roles/web/templates/parameters.yml.j2 [hunk 10]
```diff
@@ -45,3 +45,7 @@ parameters:
     # Register at https://sentry.io
     sentry_dsn:        '{{ site_sentry_dsn }}'
     sentry_env:        '{{ site_sentry_env }}'
+
+    # Algolia Configuration
+    env(ALGOLIA_APP_ID):  '{{ site_algolia_app_id }}'
```
Comment (agathver) [inline ansible/roles/web/templates/parameters.yml.j2]: This part should stay, disregard my comment in #246 
Algolia client v3 has changed the configuration interface.

File: site/src/Librecores/ProjectRepoBundle/Entity/Project.php [hunk 11]
```diff
@@ -323,6 +325,47 @@ public function __construct()
         $this->classifications = new ArrayCollection();
     }
 
+    /**
+     * Normalizes the object into an array of scalars|arrays.
+     *
+     * @param NormalizerInterface $serializer The serializer is given so that you
```
Comment (agathver) [inline site/src/Librecores/ProjectRepoBundle/Entity/Project.php]: Use `{@inheritdoc}`

File: site/src/Librecores/ProjectRepoBundle/Entity/Project.php [hunk 12]
```diff
@@ -323,6 +325,47 @@ public function __construct()
...
+     * @param string|null         $format     The format is optionally given to be able to serialize differently
+     *                                        based on different output formats
+     * @param array               $context    Options for serializing this object
+     *
+     * @return array|string|int|float|bool
+     */
+    public function normalize(NormalizerInterface $serializer, $format = null, array $context = array())
+    {
+        return [
+            'name' => $this->getName(),
```
Comment (agathver) [inline site/src/Librecores/ProjectRepoBundle/Entity/Project.php]: Will it be possible to show the project code quality score through this API? You may need to create a custom normalizer as described in https://github.com/algolia/search-bundle#using-a-custom-normalizer

**Comment by sandipbhuyan** [comment]
@imphil  and @agathver I have done some changes in this PR. That might resolve the issues here
- The custom normalizer is added. 
- use of {@inheritdoc} 
- And the configuration for aloglia-search-bundle require `env(ALGOLIA_APP_ID)` and `env(ALGOLIA_API_KEY)` in parameter.yml, So it is not wise to remove these two fields from our application

**Comment by imphil** [comment]
Now the only thing missing is the newline at the end of the file, please remove all occurrences in the diff that show 
image
in the diff view in GitHub.

**Comment by sandipbhuyan** [comment]
@imphil I have missed the new line at the end of the files now that thing is resolved.

File: site/app/config/services.yml [hunk 13]
```diff
@@ -16,3 +16,20 @@ services:
...
+
+    project_normalizer:
+        class: Librecores\ProjectRepoBundle\Serializer\Normalizer\ProjectNormalizer
+        tag: serializer.normalizer
+        public: false
+
+    organization_normalizer:
+        class: Librecores\ProjectRepoBundle\Serializer\Normalizer\OrganizationNormalizer
+        tag: serializer.normalizer
+        public: false
```
[the remark this answers is not in the document]
Comment (sandipbhuyan) [reply site/app/config/services.yml]: I have added this field as algolia search bundle configuration have described. I am checking it again for its exsistance

File: site/app/config/services.yml [hunk 14]
```diff
@@ -16,3 +16,20 @@ services:
...
+
+framework:
+  serializer: { enabled: true }
```
[the remark this answers is not in the document]
Comment (sandipbhuyan) [reply site/app/config/services.yml]: We dont have a serielizer or framework-bundle in our application thats why it requiers this feature in service configuration file. [https://github.com/algolia/search-bundle#no-serializer-service-found](https://github.com/algolia/search-bundle#no-serializer-service-found)

File: site/src/Librecores/ProjectRepoBundle/Serializer/Normalizer/ProjectNormalizer.php [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+use Symfony\Component\Serializer\Normalizer\NormalizerInterface;
+
+class ProjectNormalizer implements NormalizerInterface
+{
+    /**
+     * Normalize a Project into a set of arrays/scalars.
+     *
+     * {@inheritdoc}. This class normalizes the Projects for getting indexed in algolia
+     *
+     * @author Sandip Kumar Bhuyan <sandipbhuyan@gmail.com>
```
[the remark this answers is not in the document]
Comment (sandipbhuyan) [reply site/src/Librecores/ProjectRepoBundle/Serializer/Normalizer/ProjectNormalizer.php]: rectifying it, can I add these at the class name?

[the remark this answers is not in the document]
Comment (imphil) [reply site/src/Librecores/ProjectRepoBundle/Serializer/Normalizer/ProjectNormalizer.php]: That's what I wrote, yes.

File: site/app/config/services.yml (hunk 14 above)
[the remark this answers is not in the document]
Comment (imphil) [reply site/app/config/services.yml]: The services.yml file specifies the services that are available. Other configuration goes typically into config.yml. This is also how the Symfony docs do it: https://symfony.com/doc/3.4/serializer.html

[the remark this answers is not in the document]
Comment (sandipbhuyan) [reply site/app/config/services.yml]: If I am adding these feature in the main configuation exception is being shown by symfony
screen shot 2018-07-09 at 8 00 21 pm

File: site/app/config/services.yml (hunk 13 above)
[the remark this answers is not in the document]
Comment (imphil) [reply site/app/config/services.yml]: It's not wrong, but the old way of doing things, before autoconf and autowire existed. We have moved almost the complete codebase to these new features, and should do the same for these three new services. As I said, just remove the service registrations and things should work fine.

File: site/app/config/services.yml (hunk 14 above)
[the remark this answers is not in the document]
Comment (imphil) [reply site/app/config/services.yml]: services.yml is included in config.yml (https://github.com/librecores/librecores-web/blob/e8c0eccdeaab551925d422f3b4d77d42f307b200/site/app/config/config.yml#L4). So the result should be exactly the same -- if not then something else is wrong. Make sure to add it in the existing 'framework' key in config.yml.

[the remark this answers is not in the document]
Comment (sandipbhuyan) [reply site/app/config/services.yml]: Thanks. Resolved.

File: doc/src/devenv/tipsandtricks.rst [hunk 16]
```diff
@@ -110,3 +110,23 @@ Check the coding style of PHP code
   vm$> cd /var/www/lc/site
   vm$> ./vendor/bin/phpcs --runtime-set ignore_warnings_on_exit true -s \
     && echo You can commit: No errors found!
+
+Algolia indices configuration
+-----------------------------
+To configure Algolia in the development environment you need to specify the Application ID (site_algolia_app_id)
+and the Admin API Key (site_algolia_api_key) in the ansible/secrets/dev-vagrant.secrets.yml file.
+Clear and import the search settings from algolia
```
[the remark this answers is not in the document]
Comment (sandipbhuyan) [reply doc/src/devenv/tipsandtricks.rst]: replacing it to `you have to clear and import the search indices settings for pushing data to algolia`.

File: ansible/roles/web/templates/parameters.yml.j2 [hunk 17]
```diff
@@ -45,3 +45,7 @@ parameters:
...
+    env(ALGOLIA_API_KEY): '{{ site_algolia_api_key }}'
```
[the remark this answers is not in the document]
Comment (sandipbhuyan) [reply ansible/roles/web/templates/parameters.yml.j2]: I am adding a new line in parameters.yml.dist file too.

File: site/src/Librecores/ProjectRepoBundle/Serializer/Normalizer/ProjectNormalizer.php (hunk 15 above)
[the remark this answers is not in the document]
Comment (sandipbhuyan) [reply site/src/Librecores/ProjectRepoBundle/Serializer/Normalizer/ProjectNormalizer.php]: done

File: site/app/config/services.yml (hunk 13 above)
[the remark this answers is not in the document]
Comment (sandipbhuyan) [reply site/app/config/services.yml]: Thanks, The application is working after removing those configurations

**Comment by imphil** [comment]
squashed and merged in c784066b0a331724fefb2004285481b557d26d00. Thanks for your work on this Sandip!

**Pull request closed without merging**